storm-client/src/jvm/org/apache/storm/utils/CustomIndexArray.java

```diff
@@ -0,0 +1,174 @@
+     * @param lowerIndex Smallest (inclusive) valid index for the array. Can be +ve, -ve or 0.
+     * @param upperIndex Largest  (inclusive) valid index for the array. Can be +ve, -ve or 0. Must be > lowerIndex
+     */
+    public CustomIndexArray(int lowerIndex, int upperIndex) {
+        if (lowerIndex >= upperIndex) {
+            throw new IllegalArgumentException("lowerIndex must be < upperIndex");
+        }
+
+        this.baseIndex = lowerIndex;
+        this.elemCount = upperIndex - lowerIndex + 1;
+
+        this.elements = makeNullInitializedArray(elemCount);
+    }
+
+    private static <T> ArrayList<T> makeNullInitializedArray(int elemCount) {
+        ArrayList<T> result = new ArrayList<T>(elemCount);
+        for (int i = 0; i < elemCount; i++) {
+            result.add(null);
+        }
+        return result;
+    }
+
+    /**
+     * Initializes the array with elements from a HashTable.
+     *
+     * @param src  the source map from which to initialize
+     */
+    public CustomIndexArray(Map<Integer, T> src) {
+        Integer lowerIndex = Integer.MAX_VALUE;
+        Integer upperIndex = Integer.MIN_VALUE;
+
+        for (Map.Entry<Integer, T> entry : src.entrySet()) { // calculate smallest & largest indexes
+            Integer key =  entry.getKey();
+            if (key < lowerIndex) {
+                lowerIndex = key;
+            }
+            if (key > upperIndex) {
+                upperIndex = key;
+            }
+        }
+        this.baseIndex = lowerIndex;
+        this.elemCount = upperIndex - lowerIndex + 1;
+
+        this.elements = makeNullInitializedArray(elemCount);
+
+        for (Map.Entry<Integer, T> entry : src.entrySet()) {
+            elements.set(entry.getKey() - baseIndex, entry.getValue());
+        }
+    }
+
+    /**
+     * Get the value at the index.
+     * @throws IndexOutOfBoundsException if index is outside of bounds specified to constructor
+     */
+    public T get(int index) {
+        return elements.get(index - baseIndex);
+    }
+
+    /**
+     * Set the value at the index.
+     * @throws IndexOutOfBoundsException if index is outside of bounds specified to constructor
+     */
+    public T set(int index, T value) {
+        return elements.set(index - baseIndex, value);
+    }
+
+    /**
+     * Returns the number of elements (including null elements).
+     */
+    public int size() {
+        return elemCount;
+    }
+
+    /**
+     * Always returns true as this cannot be empty.
+     */
+    public boolean isEmpty() {
+        return false;
+    }
+
+
+    /**
+     * Check if index is valid.
+     */
+    public boolean isValidIndex(int index) {
+        int actualIndex = index - baseIndex;
+        if (actualIndex < 0  ||  actualIndex >= elemCount) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Returns the index range as a Set.
+     */
+    public Set<Integer> indices() {
```

Comment:
Nit: we are keeping the order of the indices here, but it is not documented in the javadocs, and it is not needed, as the only place this is used is for doing a set difference.

It would also be nice to indicate that this is not a cheap call so people don't abuse it calling it a lot.

Reply:
I felt its intuitive and nice to maintain the integers in order even if not absolutely necessary.